nntrainer/include/layer.h

```diff
@@ -479,20 +473,34 @@ class Layer {
   bool trainable;
 
   /**
-   * @brief     Gradient for the weights in this layer
-   * @note      The order of gradients should match the order in weights
+   * @brief     reserve memory for @a params and set @a param_size
+   * @exception std::invalid_argument when param_size is already set and
+   * shouldn't be changed again.
    */
-  std::vector<std::reference_wrapper<Tensor>> gradients;
+  void setParamSize(unsigned int psize) {
+
+    // @note Need opinion about this
+    // if (param_size > 0) {
+    //   throw std::invalid_argument("param size can't be set once it is set");
+    // }
+
+    param_size = psize;
+    params = std::shared_ptr<UpdatableParam>(
+      new UpdatableParam[psize], std::default_delete<UpdatableParam[]>());
+  }
 
   /**
-   * @brief     weights in this layer
-   * @note      The weights are combined with their corresponding bias
-   *            For example- with W0, W1, B0 and B1, weights would be of format
-   *            {W0, B0, W1, B1}.
+   * @brief     get data alias at param position.
+   * @exception std::out_of_range for index out of range
    */
-  std::vector<std::reference_wrapper<Tensor>> weights;
+  UpdatableParam &paramsAt(const unsigned int position) {
+    if (position >= param_size) {
+      throw std::out_of_range("index out of range");
+    }
+
+    return params.get()[position];
+  }
 
-private:
```

Comment:
why move all of these to protected?

Reply:
Oh, I just put it back. Don't know why yesterday me deleted that